File: aisd/lista4/zad1/bst.py
import random
import sys
import time

big_tree = False

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.parent = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

        # counters for current operation
        self.current_comparisons = 0
        self.current_pointer_ops = 0

        # stats collection
        self.insert_comparisons_log = []
        self.insert_pointer_ops_log = []
        self.insert_heights_log = []

        self.delete_comparisons_log = []
        self.delete_pointer_ops_log = []
        self.delete_heights_log = []



    def reset_counters(self):
        self.current_comparisons = 0
        self.current_pointer_ops = 0

    def log_operation(self, op):
        if op == "delete":
            self.delete_comparisons_log.append(self.current_comparisons)
            self.delete_pointer_ops_log.append(self.current_pointer_ops)
            self.delete_heights_log.append(self.tree_height())
        elif op == "insert":
            self.insert_comparisons_log.append(self.current_comparisons)
            self.insert_pointer_ops_log.append(self.current_pointer_ops)
            self.insert_heights_log.append(self.tree_height())

    def search(self, node, key):
        current = node
        while current is not None:
            self.current_comparisons += 1
            if key == current.key:
                return current
            self.current_comparisons += 1
            if key < current.key:
                current = current.left
                self.current_pointer_ops += 1
            else:
                current = current.right
                self.current_pointer_ops += 1
        return None

    def insert_node(self, z):
        self.reset_counters()
        y = None
        x = self.root
        self.current_pointer_ops += 1

        while x is not None:
            y = x
            self.current_comparisons += 1
            if z.key < x.key:
                x = x.left
                self.current_pointer_ops += 1
            else:
                x = x.right
                self.current_pointer_ops += 1

        z.parent = y
        self.current_pointer_ops += 1
        if y is None:
            self.root = z          # tree was empty
            self.current_pointer_ops += 1
        elif z.key < y.key:
            y.left = z
            self.current_pointer_ops += 1
            self.current_comparisons += 1
        else:
            y.right = z
            self.current_pointer_ops += 1
            self.current_comparisons += 1

        self.log_operation("insert")

        # replaces one subtree as a child of its parent with another subtree, needed for delete
    def transplant(self, u, v):
        self.current_pointer_ops += 1
        if u.parent is None:
            self.root = v
            self.current_pointer_ops += 1  # assign root
        elif u == u.parent.left:
            u.parent.left = v
            self.current_pointer_ops += 4  # condition plus previous line
        else:
            u.parent.right = v
            self.current_pointer_ops += 4  # we need to count condition also

        if v is not None:
            v.parent = u.parent
            self.current_pointer_ops += 2

    def tree_minimum(self, x):
        while x.left is not None:
            x = x.left
            self.current_pointer_ops += 2
        return x

    def tree_maximum(self, x):
        while x.right is not None:
            x = x.right
            self.current_pointer_ops += 2
        return x

    def delete_node(self, val_to_delete):
        self.reset_counters()

        z = self.search(self.root, val_to_delete)

        self.current_pointer_ops += 1
        if z.left is None:
            self.transplant(z, z.right)
            self.current_pointer_ops += 1
        elif z.right is None:
            self.transplant(z, z.left)
            self.current_pointer_ops += 2
        else:
            y = self.tree_minimum(z.right)
            self.current_pointer_ops += 2

            self.current_pointer_ops += 1
            if y.parent != z:
                self.transplant(y, y.right)
                y.right = z.right
                self.current_pointer_ops += 4
                if y.right is not None:
                    y.right.parent = y
                    self.current_pointer_ops += 2

            self.transplant(z, y)
            y.left = z.left
            self.current_pointer_ops += 3
            if y.left is not None:
                y.left.parent = y
                self.current_pointer_ops += 2

        self.log_operation("delete")

    def tree_height(self):
        if not self.root:
            return 0
        max_height = 0
        stack = [(self.root, 1)]
        while stack:
            node, depth = stack.pop()
            if node:
                max_height = max(max_height, depth)
                stack.append((node.left, depth + 1))
                stack.append((node.right, depth + 1))
        return max_height

    def print_BST(self):
        if self.root is not None:
            print_tree(self.root, val='key')
        else:
            print("(empty tree)")

def print_tree(root, val="val", left="left", right="right"):
    def display(root, val=val, left=left, right=right):
        # no children
        if getattr(root, right) is None and getattr(root, left) is None:
            line = '%s' % getattr(root, val)
            width = len(line)
            height = 1
            middle = width // 2
            return [line], width, height, middle

        # only left child
        if getattr(root, right) is None:
            lines, n, p, x = display(getattr(root, left))       # recurrent in left child
            s = '%s' % getattr(root, val)
            u = len(s)
            first_line = (x + 1) * ' ' + (n - x - 1) * '_' + s
            second_line = x * ' ' + '/' + (n - x - 1 + u) * ' '
            shifted_lines = [line + u * ' ' for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, n + u // 2

        # only right child
        if getattr(root, left) is None:
            lines, n, p, x = display(getattr(root, right))      # recurrent in right child
            s = '%s' % getattr(root, val)                       # value of current node
            u = len(s)                                          # length of value
            first_line = s + x * '_' + (n - x) * ' '
            second_line = (u + x) * ' ' + '\\' + (n - x - 1) * ' '
            shifted_lines = [u * ' ' + line for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, u // 2

        left_lines, n, p, x = display(getattr(root, left))
        right_lines, m, q, y = display(getattr(root, right))
        s = '%s' % getattr(root, val)
        u = len(s)
        first_line = (x + 1) * ' ' + (n - x - 1) * '_' + s + y * '_' + (m - y) * ' '
        second_line = x * ' ' + '/' + (n - x - 1 + u + y) * ' ' + '\\' + (m - y - 1) * ' '
        # fill left and right tree
        if p < q:
            left_lines += [n * ' '] * (q - p)
        elif q < p:
            right_lines += [m * ' '] * (p - q)

        # merge two trees: left and right
        zipped_lines = zip(left_lines, right_lines)
        lines = [first_line, second_line] + [a + u * ' ' + b for a, b in zipped_lines]
        return lines, n + m + u, max(p, q) + 2, n + u // 2

    # calling display and writing a tree
    lines, *_ = display(root, val, left, right)
    for line in lines:
        print(line)

def ascending_insert(tree, n):
    for i in range(1, n + 1):
        node_to_insert = Node(i)
        tree.insert_node(node_to_insert)
        if not big_tree:
            print()
            print("INSERT: ", i)
            tree.print_BST()
            print("Height: ", tree.tree_height())

def random_insert(tree, n):
    keys = list(range(1, n + 1))    # creating random permutation
    random.shuffle(keys)
    for key in keys:
        node_to_insert = Node(key)
        tree.insert_node(node_to_insert)
        if not big_tree:
            print()
            print("INSERT: ", key)
            tree.print_BST()
            print("Height: ", tree.tree_height())

def random_delete(tree, n):
    keys = list(range(1, n + 1))
    random.shuffle(keys)

    for key in keys:
        tree.delete_node(key)
        if not big_tree:
            print()
            print("DELETE: ", key)
            tree.print_BST()
            print("Height: ", tree.tree_height())

def metrics_cost(log):
    if not log:
        return 0, 0
    return sum(log)/len(log)

if __name__ == "__main__":

    if(len(sys.argv) != 2):
        print("Wrong number of arguments")
        sys.exit(1)

    try:
        n = int(sys.argv[1])
    except ValueError:
        print("Error: n has to be an integer")
        sys.exit(1)

    if n > 30:
        big_tree = True

    bst = BinarySearchTree()

    start = time.time()

    print("----1 case: ascending_insert and random_delete for n =", n, "----")
    ascending_insert(bst, n)
    print("Ascending insert average cost comparisons: ", metrics_cost(bst.insert_comparisons_log))
    print("Ascending insert average cost pointer: ", metrics_cost(bst.insert_pointer_ops_log))
    print("Ascending insert average height: ", metrics_cost(bst.insert_heights_log))

    print("Ascending insert max comparisons: ", max(bst.insert_comparisons_log))
    print("Ascending insert max pointer: ", max(bst.insert_pointer_ops_log))
    print("Ascending insert max height: ", max(bst.insert_heights_log))

    random_delete(bst, n)
    print("Random delete average cost comparisons: ", metrics_cost(bst.delete_comparisons_log))
    print("Random delete average cost pointer: ", metrics_cost(bst.delete_pointer_ops_log))
    print("Random delete average height: ", metrics_cost(bst.delete_heights_log))

    print("Random delete max comparisons: ", max(bst.delete_comparisons_log))
    print("Random delete max pointer: ", max(bst.delete_pointer_ops_log))
    print("Random delete max height: ", max(bst.delete_heights_log))
    print()

    bst2 = BinarySearchTree()

    print("----2 case: random_insert and random_delete for n =", n, "----")
    random_insert(bst2, n)
    print("Random insert average cost comparisons: ", metrics_cost(bst2.insert_comparisons_log))
    print("Random insert average cost pointer: ", metrics_cost(bst2.insert_pointer_ops_log))
    print("Random insert average height: ", metrics_cost(bst2.insert_heights_log))

    print("Random insert max comparisons: ", max(bst2.insert_comparisons_log))
    print("Random insert max pointer: ", max(bst2.insert_pointer_ops_log))
    print("Random insert max height: ", max(bst2.insert_heights_log))

    random_delete(bst2, n)
    print("Random delete average cost comparisons: ", metrics_cost(bst2.delete_comparisons_log))
    print("Random delete average cost pointer: ", metrics_cost(bst2.delete_pointer_ops_log))
    print("Random delete average height: ", metrics_cost(bst2.delete_heights_log))

    print("Random delete max comparisons: ", max(bst2.delete_comparisons_log))
    print("Random delete max pointer: ", max(bst2.delete_pointer_ops_log))
    print("Random delete max height: ", max(bst2.delete_heights_log))
    end = time.time()

    print("Time elapsed: ", end - start)

File: aisd/lista4cpp/zad1/bst.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <algorithm>
#include <random>
#include <chrono>
#include <numeric>
#include <cstdlib>

using namespace std;
enum Operation {INSERT, DELETE};

bool big_tree = false;

class Node {
  public:
    int value;
    Node* left;
    Node* right;
    Node* parent;
    int height;

  // normal constructor
  Node(int val)
      : value(val), left(getNIL()), right(getNIL()), parent(getNIL()), height(1) {}

  // constructor for NIL
  Node(int val, bool is_nil)
    : value(val), left(nullptr), right(nullptr), parent(nullptr), height(0) {}

  static Node* getNIL() {
    static Node* NIL = []() {
      Node* nil = new Node(0, true);  // second constructor
      nil->left = nil;
      nil->right = nil;
      nil->parent = nil;
      return nil;
    }();
    return NIL;
  }
};

class BinarySearchTree {
  public:
    Node* root;

    // counters for current operation
    int current_comparisons = 0;
    int current_pointer_ops = 0;

    // stats collection
    vector<int> insert_comparisons;
    vector<int> insert_pointer_ops;
    vector<int> insert_heights;

    vector<int> delete_comparisons;
    vector<int> delete_pointer_ops;
    vector<int> delete_heights;

  BinarySearchTree() {
      root = Node::getNIL();
  }

  void reset_counters() {
    current_comparisons = 0;
    current_pointer_ops = 0;
  }

  int node_height(Node* node) {
    if (node == Node::getNIL()) return 0;
    return node->height;
  }

  void update_height(Node* node) {
    if (node != Node::getNIL()) {
      node->height = 1 + max(node_height(node->left), node_height(node->right));
    }
  }
  void update_height_upwards(Node* node) {
    while (node != Node::getNIL()) {
      update_height(node);
      node = node->parent;
    }
  }

  int tree_height() {

    return root->height;
  }

    void log_operation(Operation op) {
      if(op == INSERT) {
        insert_comparisons.push_back(current_comparisons);
        insert_pointer_ops.push_back(current_pointer_ops);
        insert_heights.push_back(tree_height());
      }
      else if(op == DELETE) {
        delete_comparisons.push_back(current_comparisons);
        delete_pointer_ops.push_back(current_pointer_ops);
        delete_heights.push_back(tree_height());
      }
    }

    Node* search(Node* node, int key) {
      while (node != Node::getNIL()) {
        current_comparisons++;
        if (key == node->value) {
          return node;
        }
        current_comparisons++;
        current_pointer_ops++;
        if (key < node->value) {
          node = node->left;
        } else {
          node = node->right;
        }
      }
      return Node::getNIL();
    }

  void insert_node(Node* z) {
    reset_counters();

    Node* y = Node::getNIL();
    Node* x = root;
    current_pointer_ops++;

    while (x != Node::getNIL()) {
      y = x;
      current_comparisons++;
      current_pointer_ops++;
      if (z->value < x->value) {
        x = x->left;
      } else {
        x = x->right;
      }
    }
    z->parent = y;
    current_pointer_ops++;

    if (y == Node::getNIL()) {
      root = z;              // tree was empty
      current_pointer_ops++;
    } else if (z->value < y->value) {
      y->left = z;
      current_pointer_ops++;
      current_comparisons++;
    } else {
      y->right = z;
      current_pointer_ops++;
      current_comparisons++;
    }

    // go up and update heights in nodes
    update_height_upwards(z->parent);

    log_operation(INSERT);
  }

  void transplant(Node* u, Node* v) {
    current_pointer_ops++;
    if (u->parent == Node::getNIL()) {
      root = v;
      current_pointer_ops++;
    }
    else if (u == u->parent->left) {
      u->parent->left = v;
      current_pointer_ops += 4;
    }
    else {
      u->parent->right = v;
      current_pointer_ops += 4;
    }

    if (v != Node::getNIL()) {
      v->parent = u->parent;
      current_pointer_ops += 2;
    }
  }

  Node* tree_minimum(Node* x) {
    while (x->left != Node::getNIL()) {
      x = x->left;
      current_pointer_ops += 2;
    }
    return x;
  }

  Node* tree_maximum(Node* x) {
    while (x->right != Node::getNIL()) {
      x = x->right;
      current_pointer_ops += 2;
    }
    return x;
  }

  void delete_node(int val_to_delete) {
    reset_counters();

    Node* z = search(root, val_to_delete);    // node which we found

    Node* update_start = Node::getNIL();

    current_pointer_ops++;
    if (z->left == Node::getNIL()) {
      update_start = z->parent;
      transplant(z, z->right);
      current_pointer_ops++;
    } else if (z->right == Node::getNIL()) {
      update_start = z->parent;
      transplant(z, z->left);
      current_pointer_ops += 2;
    } else {
      Node* y = tree_minimum(z->right);
      current_pointer_ops += 2;

      update_start = y;

      current_pointer_ops++;
      if (y->parent != z) {
        transplant(y, y->right);
        y->right = z->right;
        current_pointer_ops += 4;
        if (y->right != Node::getNIL()) {
          y->right->parent = y;
          current_pointer_ops += 2;
        }
      }
      transplant(z, y);
      y->left = z->left;
      current_pointer_ops += 3;
      if (y->left != Node::getNIL()) {
        y->left->parent = y;
        current_pointer_ops += 2;
      }
    }
    update_height_upwards(update_start);
    log_operation(DELETE);
  }
};

string to_string_node(Node* node) {
    if (node == Node::getNIL()) return "NIL";
    stringstream ss;
    ss << node->value;
    return ss.str();
}

vector<string> display(Node* root, int& width, int& height, int& middle) {
  if (root == Node::getNIL()) {
      width = 0;
      height = 0;
      middle = 0;
      return {};
  }

  string s = to_string_node(root);
  int u = s.length();

  // No children
  if (root->left == Node::getNIL() && root->right == Node::getNIL()) {
      width = u;
      height = 1;
      middle = u / 2;
      return { s };
  }

  // Only left child
  if (root->right == Node::getNIL()) {
      int n, p, x;
      vector<string> left_lines = display(root->left, n, p, x);
      string first_line = string(x + 1, ' ') + string(n - x - 1, '_') + s;
      string second_line = string(x, ' ') + '/' + string(n - x - 1 + u, ' ');

      vector<string> shifted_lines;
      for (const string& line : left_lines)
          shifted_lines.push_back(line + string(u, ' '));

      width = n + u;
      height = p + 2;
      middle = n + u / 2;
      vector<string> result = { first_line, second_line };
      result.insert(result.end(), shifted_lines.begin(), shifted_lines.end());
      return result;
  }

  // Only right child
  if (root->left == Node::getNIL()) {
      int m, q, y;
      vector<string> right_lines = display(root->right, m, q, y);
      string first_line = s + string(y, '_') + string(m - y, ' ');
      string second_line = string(u + y, ' ') + '\\' + string(m - y - 1, ' ');

      vector<string> shifted_lines;
      for (const string& line : right_lines)
          shifted_lines.push_back(string(u, ' ') + line);

      width = m + u;
      height = q + 2;
      middle = u / 2;
      vector<string> result = { first_line, second_line };
      result.insert(result.end(), shifted_lines.begin(), shifted_lines.end());
      return result;
  }

  // Both children
  int n, p, x;
  int m, q, y;
  vector<string> left_lines = display(root->left, n, p, x);
  vector<string> right_lines = display(root->right, m, q, y);
  string first_line = string(x + 1, ' ') + string(n - x - 1, '_') + s + string(y, '_') + string(m - y, ' ');
  string second_line = string(x, ' ') + '/' + string(n - x - 1 + u + y, ' ') + '\\' + string(m - y - 1, ' ');

  if (p < q) left_lines.resize(q, string(n, ' '));
  if (q < p) right_lines.resize(p, string(m, ' '));

  vector<string> lines;
  for (size_t i = 0; i < left_lines.size(); ++i)
      lines.push_back(left_lines[i] + string(u, ' ') + right_lines[i]);

  width = n + m + u;
  height = max(p, q) + 2;
  middle = n + u / 2;
  vector<string> result;
  result.push_back(first_line);
  result.push_back(second_line);
  result.insert(result.end(), lines.begin(), lines.end());
  return result;
}

// main func to printing tree
void print_BST(Node* root) {
    if (root == Node::getNIL()) {
        cout << "(empty tree)" << endl;
        return;
    }

    int width, height, middle;
    vector<string> lines = display(root, width, height, middle);
    for (const string& line : lines)
        cout << line << endl;
}

void ascending_insert(BinarySearchTree* tree, int n) {
  for(int i = 1; i <= n; i++) {
    Node* node_to_insert = new Node(i);
    tree->insert_node(node_to_insert);
    if (!big_tree) {
      cout << endl;
      cout << "INSERT: " << i << endl;
      print_BST(tree->root);
      cout << "Height: " << tree->tree_height() << endl;
    }
  }
}

// time-based seed
unsigned seed = chrono::system_clock::now().time_since_epoch().count();

void random_insert(BinarySearchTree* tree, int n) {
  vector<int> keys;
  for (int i = 1; i <= n; ++i) {
    keys.push_back(i);
  }
  shuffle(keys.begin(), keys.end(), default_random_engine(seed));

  for (int key : keys) {
    Node* node_to_insert = new Node(key);
    tree->insert_node(node_to_insert);
    if (!big_tree) {
      cout << "\nINSERT: " << key << endl;
      print_BST(tree->root);
      cout << "Height: " << tree->tree_height() << endl;
    }
  }
}
void random_delete(BinarySearchTree* tree, int n) {
  std::vector<int> keys;
  for (int i = 1; i <= n; ++i) {
    keys.push_back(i);
  }
  std::shuffle(keys.begin(), keys.end(), default_random_engine(seed));

  for (int key : keys) {
    tree->delete_node(key);
    if (!big_tree) {
      cout << "\nDELETE: " << key << endl;
      print_BST(tree->root);
      cout << "Height: " << tree->tree_height() << endl;
    }
  }
}

int metrics_cost(vector<int> log) {
  if (log.size() == 0) return 0;
  int sum = accumulate(log.begin(), log.end(), 0);
  return sum / log.size();
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cout << "Wrong number of arguments\n";
        return 1;
    }

    int n;
    try {
        n = stoi(argv[1]);
    } catch (...) {
        cout << "Error: n has to be an integer\n";
        return 1;
    }

    if (n > 30) {
        big_tree = true;
    }

    BinarySearchTree bst;

    auto start = chrono::steady_clock::now();

    cout << "----1 case: ascending_insert and random_delete for n = " << n << "----\n";

    ascending_insert(&bst, n);

    cout << "Ascending insert average cost comparisons: " << metrics_cost(bst.insert_comparisons) << "\n";
    cout << "Ascending insert average cost pointer: " << metrics_cost(bst.insert_pointer_ops) << "\n";
    cout << "Ascending insert average height: " << metrics_cost(bst.insert_heights) << "\n";

    if (!bst.insert_comparisons.empty())
        cout << "Ascending insert max comparisons: " << *max_element(bst.insert_comparisons.begin(), bst.insert_comparisons.end()) << "\n";
    if (!bst.insert_pointer_ops.empty())
        cout << "Ascending insert max pointer: " << *max_element(bst.insert_pointer_ops.begin(), bst.insert_pointer_ops.end()) << "\n";
    if (!bst.insert_heights.empty())
        cout << "Ascending insert max height: " << *max_element(bst.insert_heights.begin(), bst.insert_heights.end()) << "\n";

    random_delete(&bst, n);

    cout << "Random delete after ascending insert average cost comparisons: " << metrics_cost(bst.delete_comparisons) << "\n";
    cout << "Random delete after ascending insert average cost pointer: " << metrics_cost(bst.delete_pointer_ops) << "\n";
    cout << "Random delete after ascending insert average height: " << metrics_cost(bst.delete_heights) << "\n";

    if (!bst.delete_comparisons.empty())
        cout << "Random delete after ascending insert max comparisons: " << *max_element(bst.delete_comparisons.begin(), bst.delete_comparisons.end()) << "\n";
    if (!bst.delete_pointer_ops.empty())
        cout << "Random delete after ascending insert max pointer: " << *max_element(bst.delete_pointer_ops.begin(), bst.delete_pointer_ops.end()) << "\n";
    if (!bst.delete_heights.empty())
        cout << "Random delete after ascending insert max height: " << *max_element(bst.delete_heights.begin(), bst.delete_heights.end()) << "\n";

    cout << "\n";

    BinarySearchTree bst2;

    cout << "----2 case: random_insert and random_delete for n = " << n << "----\n";

    random_insert(&bst2, n);

    cout << "Random insert average cost comparisons: " << metrics_cost(bst2.insert_comparisons) << "\n";
    cout << "Random insert average cost pointer: " << metrics_cost(bst2.insert_pointer_ops) << "\n";
    cout << "Random insert average height: " << metrics_cost(bst2.insert_heights) << "\n";

    if (!bst2.insert_comparisons.empty())
        cout << "Random insert max comparisons: " << *max_element(bst2.insert_comparisons.begin(), bst2.insert_comparisons.end()) << "\n";
    if (!bst2.insert_pointer_ops.empty())
        cout << "Random insert max pointer: " << *max_element(bst2.insert_pointer_ops.begin(), bst2.insert_pointer_ops.end()) << "\n";
    if (!bst2.insert_heights.empty())
        cout << "Random insert max height: " << *max_element(bst2.insert_heights.begin(), bst2.insert_heights.end()) << "\n";

    random_delete(&bst2, n);

    cout << "Random delete after random insert average cost comparisons: " << metrics_cost(bst2.delete_comparisons) << "\n";
    cout << "Random delete after random insert average cost pointer: " << metrics_cost(bst2.delete_pointer_ops) << "\n";
    cout << "Random delete after random insert average height: " << metrics_cost(bst2.delete_heights) << "\n";

    if (!bst2.delete_comparisons.empty())
        cout << "Random delete after random insert max comparisons: " << *max_element(bst2.delete_comparisons.begin(), bst2.delete_comparisons.end()) << "\n";
    if (!bst2.delete_pointer_ops.empty())
        cout << "Random delete after random insert max pointer: " << *max_element(bst2.delete_pointer_ops.begin(), bst2.delete_pointer_ops.end()) << "\n";
    if (!bst2.delete_heights.empty())
        cout << "Random delete after random insert max height: " << *max_element(bst2.delete_heights.begin(), bst2.delete_heights.end()) << "\n";

    auto end = chrono::steady_clock::now();
    auto elapsed = chrono::duration_cast<chrono::milliseconds>(end - start).count();

    cout << "Time elapsed: " << elapsed / 1000.0 << " seconds\n";

    return 0;
}

